### PR TITLE
Replacing uniform with uniform_philox , and incrementing PRNG state

### LINKED ISSUE #114: Add `torch.nn.Dropout` recomputation support during the backward pass to Thunder
## 🚀 Feature

I would like to have Thunder save the seed and offset from random number generation to allow for the recomputation of Dropout in the backward pass.

There are two pieces needed to make it work:
* Support stateless (deterministic) PRNG. This is done with `thunder.prims.uniform_philox`.
* Trace transform to query PyTorch's PRNG state before each `uniform` call, replacing `uniform` with `uniform_philox` , and incrementing PRNG state properly. This is not implemented.

### Motivation

Multihead Attention modules in LLMs often use dropout where the memory used is the square of the sequence length.

cc @apaz-cli

### LINKED ISSUE #231: If saved_for_backward returns NumberProxy, the value is taken from compile time, not runtime
## 🐛 Bug

When I add an operator that returns numbers, the values in the saved_for_backward are the compile time value defined in `Symbol.meta`, not the real value computed at runtime.

An example trace is like:
```py
@torch.no_grad()
@no_autocast()
def augmented_forward_fn(a):
  # a: "cuda:0 f32[2, 2]"
  t2 = get_rng_state_prim_impl(None, devices.Device("cuda:0"))  # t2: "cpu ui8[16]"
  (i3, i4) = unpack_rng_state_prim_impl(t2)
  del t2
  [t1] = nvFusion0(a, i3, i4)
    # t0 = prims.uniform_philox((2, 2), 0.0, 1.0, device=devices.Device("cuda:0"), dtype=dtypes.float32, seed=i3, offset=i4)  # t0: "cuda:0 f32[2, 2]"
    # t1 = prims.mul(t0, a)  # t1: "cuda:0 f32[2, 2]"
  i6 = operator.add(i4, 4)  # i6: "int 11"
    # i6 = prims.add(i4, 4)  # i6: "int 11"
  del i4
  t7 = pack_rng_state_prim_impl(i3, i6)  # t7: "cpu ui8[16]"
  del i3, i6
  set_rng_state_prim_impl(t7, devices.Device("cuda:0"))
  del t7
  ######### i3, i4 is not passed to backward, but the constant value 7 in the meta function
  return {'output': t1, 'flat_args': [a], 'flat_output': (t1,)}, ((), (7, 7))  

@torch.no_grad()
@no_autocast()
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  _, C1, = saved_for_backward
  clear_collection(saved_for_backward)
  del saved_for_backward
  t2, = cotangents
  clear_collection(cotangents)
  del cotangents
  i3, i4, = C1
  clear_collection(C1)
  del C1
  [t12] = nvFusion0(i3, i4, t2)
    # t0 = prims.uniform_philox((2, 2), 0.0, 1.0, device=devices.Device("cuda:0"), dtype=dtypes.float32, seed=i3, offset=i4)  # t0: "cuda:0 f32[2, 2]"
    # t12 = prims.mul(t0, t2)  # t12: "cuda:0 f32[2, 2]"
  del i3, i4, t2
  return (t12,)
```

### To Reproduce
Reproduction is based on branch `uniform_rng`

```py
import torch
import thunder

def func(a):
    b = thunder.torch.uniform_like(a, device=a.device, dtype=a.dtype)
    return b*a

a = torch.randn(2, 2, device="cuda", requires_grad=True)

jfunc = thunder.jit(func)
out = jfunc(a)

print(thunder.last_traces(jfunc)[-1])
print(thunder.last_backward_traces(jfunc)[-1])
```

### PR DESCRIPTION
**Pull request by kiya00**
This PR is one step towards recomputing dropout in backward, the steps are:

- [x] Add the prim operators to query/update the default CUDA RNG state [PR](https://github.com/Lightning-AI/lightning-thunder/pull/264/)
- [ ] Fix #231
- [x] Add transformation to replacing uniform with uniform_philox (this PR)
- [ ] Make sure the rematerialization works properly for Dropout

Ref #114

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
Let's split `replace_uniform` into a separate PR.
Primitives for getting and setting the state can be tested without `replace_uniform` and having https://github.com/Lightning-AI/lightning-thunder/issues/231 fixed is not required for these primitives.

**Comment by kiya00** [comment]
Hi @mruberry @jjsjann123 @IvanYashchuk , I modified it according our design review discussion, could you take a look?

**Comment by IvanYashchuk** [comment]
> merged 9 commits into [rng_update_prims](https://github.com/Lightning-AI/lightning-thunder/tree/rng_update_prims) from [uniform_rng](https://github.com/Lightning-AI/lightning-thunder/tree/uniform_rng)

We had a manual stack here 😃

**Comment by t-vi** [comment]
Yeah, I'll learn to read...
